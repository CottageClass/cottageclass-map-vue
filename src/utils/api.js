import Vue from 'vue'
import camelcaseKeys from 'camelcase-keys'
import * as Token from './tokens.js'
import normalize from 'json-api-normalizer';

var moment = require('moment');

/*
 * PROXY SESSIONS
 */
export function initProxySession(currentUserId, receiverId, requestMessage, acknowledgmentMessage) {
  console.log("INITIATING PROXY WITH users " + currentUserId + ", " + receiverId)
  let postData = {
    twilioSession: {
      requestMessage: requestMessage,
      acknowledgmentMessage: acknowledgmentMessage,
    }
  }
  return Vue.axios.post(
    `${process.env.BASE_URL_API}/users/${receiverId}/proxy_sessions`,
    postData,
  ).then(res => {
    console.log("proxy session init SUCCESS, returning proxy phone number for receiver")
    console.log(res)
    // return proxy number for receiver
    return res.data.data.attributes.proxyIdentifierReceiver
  }).catch(err => {
    console.log("proxy session init FAILURE")
    console.log(err)
    throw err
  })
}

/*
 * USERS
 */

function createPersonObject (personInApi, availableChildren = []) {
  var p = personInApi.attributes
  let hasAllRequiredFields = function () {
    if (p.phone && p.latitude && p.longitude && p.network_code) {
      return true
    } else {
      return false
    }
  }
  let createChildrenList = function () {
    let parseChild = function (child) {
      return {
        age: child.attributes.age
      }
    }
    return availableChildren.filter(child => child.attributes.parent_id == personInApi.id).map(parseChild)
    // make sure this is an array.
  }

  let activities = p.activities || []
  activities = activities.map(activity => activity.replace(/_/g, " "))

  return {
    agreeTos: p.agree_tos,
    id: personInApi.id,
    firstName: p.first_name,
    lastInitial: p.last_name && p.last_name[0],
    avatar: p.avatar,
    activities: activities,
    availableMornings: p.available_mornings,
    availableEvenings: p.available_evenings,
    availableAfternoons: p.available_afternoons,
    availableWeekends: p.available_weekends,
    location: {
      lat: parseFloat(p.fuzzy_latitude),
      lng: parseFloat(p.fuzzy_longitude)
    },
    // todo: add these once API has them
    title: "",
    employer: "",
    backgroundCheck: false,
    facebookId: p.facebook_id,
    facebookMapIcon: 'https://graph.facebook.com/' + p.facebook_id + '/picture?width=30',
    // todo: add children now somehow
    children: createChildrenList(),
    // todo: add these once I have them
    title: "",
    employer: "",
    verified: p.verified,
    phone: p.phone,
    networkCode: p.network_code,
    dateCreated: p.date_created,
    hasAllRequiredFields: hasAllRequiredFields(),
    blurb: p.profile_blurb
  }
}

// parses responseData into peopleArray and childrenArray
function createPeopleObject (responseData) {
  let peopleDataArray = responseData.data
  let included = responseData.included || []
  let childrenArray = included.filter(obj => obj.type === "child")
  return peopleDataArray.map(personInApi => createPersonObject(personInApi, childrenArray))
}

export function fetchUsersInNetwork(networkId) {
  return Vue.axios.get(
    `${process.env.BASE_URL_API}/networks/${networkId}/users`
  ).then(res => {
    console.log("FETCH USERS IN NETWORK SUCCESS")
    console.log(res.data)
    return createPeopleObject(res.data)
  }).catch(err => {
    console.log("FETCH USERS IN NETWORK FAILURE")
    console.log(err.errors)
    throw err
  })
}

export function fetchUsersWhoHaveMadeInquiries(currentUserId) {
  return Vue.axios.get(
    `${process.env.BASE_URL_API}/users/${currentUserId}/inquiries`
    ).then(res => {
      console.log("FETCH USERS WHO HAVE MADE INQUIRIES SUCCESS")
      console.log(res.data)
          return createPeopleObject(res.data)
        }).catch(err => {
          console.log("FETCH USERS WHO HAVE MADE INQUIRIES FAILURE")
          console.log(err.errors)
          throw err
        })
}

export function fetchCurrentUser(userId) {
  return Vue.axios.get(
    `${process.env.BASE_URL_API}/users/${userId}`
  ).then(res => {
    console.log("FETCH CURRENT USER SUCCESS")
    console.log(createPersonObject(res.data.data))
    return createPersonObject(res.data.data)
  }).catch(err => {
    console.log("FETCH CURRENT USER FAILURE")
    console.log(err.errors)
    throw err
  })
}


// same as above but using 'normalize' json normalizer to correctly extract children
export function fetchCurrentUserNew(userId) {
  return Vue.axios.get(
    `${process.env.BASE_URL_API}/users/${userId}`
  ).then(res => {
    console.log("FETCH CURRENT USER SUCCESS")
    console.log(res)
    let normalizedData = normalize(res.data)
    let user = normalizedData.user[userId].attributes
    user.hasAllRequiredFields = user.phone && user.latitude && user.longitude && user.networkCode
    let childrenById = normalizedData.child
    let childIds = Object.keys(childrenById)
    let generateChild = function (aChildId) {
      let child = childrenById[aChildId].attributes
      child.id = aChildId
      child.firstName = capitalize(child.firstName)
      return child
    } 
    user.children = childIds.map(generateChild)
    return user
  }).catch(err => {
    console.log("FETCH CURRENT USER FAILURE")
    console.log(err.errors)
    throw err
  })
}

// backend requires user to be an admin
export function fetchAllUsers() {
  return Vue.axios.get(
    `${process.env.BASE_URL_API}/users`
  ).then(res => {
    console.log("FETCH ALL USERS SUCCESS")
    console.log(createPeopleObject(res.data))
    return createPeopleObject(res.data)
  }).catch(err => {
    console.log("FETCH ALL USERS FAILURE")
    console.log(err.errors)
    throw err
  })
}

/*
 * MESSAGES
 */

function createMessagesObject(msgListFromApi) {
  return msgListFromApi.map(createMessageObject)
}

function createMessageObject(msgFromApi) {
  return {
    id: msgFromApi.id,
    ...camelcaseKeys(msgFromApi.attributes, {deep: true})
  }
}

export function fetchMessagesForUserPair(participantId1, participantId2) {
  return Vue.axios.get(
    `${process.env.BASE_URL_API}/users/${participantId1}/messages/${participantId2}`
  ).then(res => {
    console.log("FETCH Messages for User Pair SUCCESS")
    console.log(createMessagesObject(res.data.data))
    return createMessagesObject(res.data.data)
  }).catch(err => {
    console.log("FETCH Messages for User Pair FAILURE")
    console.log(err)
    console.log(err.errors)
    throw err
  })
}

/*
 * NOTIFICATIONS
 */

export function submitNotification(participantId, notificationBodyText) {
  let notificationData = {
  'notification': {
    'body': notificationBodyText
  }
}
  return Vue.axios.post(
    `${process.env.BASE_URL_API}/api/users/${participantId}/notifications/`, notificationData
  ).then(res => {
    console.log("NOTIFICATION submission SUCCESS")
    return res
  }).catch(err => {
    console.log("NOTIFICATION submission FAILURE")
    console.log(err)
    throw err
  })
}


/*
 * EVENTS
 */

export function fetchEvents(scopeOrEventId) {
  return Vue.axios.get(
    `${process.env.BASE_URL_API}/api/events/${scopeOrEventId || ''}`
    ).then(res => {
      console.log("FETCH EVENTS SUCCESS")
      console.log(res.data)
      return Object.values(normalize(res.data).event).map(obj => {
      var e = obj.attributes
      e["id"] = obj.id
      e.hostFirstName = capitalize(e.hostFirstName)
      e.hostFuzzyLatitude = parseFloat(e.hostFuzzyLatitude)
      e.hostFuzzyLongitude = parseFloat(e.hostFuzzyLongitude)
      e.activityName = e.activityNames.length > 0 && e.activityNames[0]
      e.food = e.foods.length > 0 && e.foods[0]
      return e
      })
        }).catch(err => {
          console.log("FETCH EVENTS FAILURE")
          console.log(err.errors)
          throw err
        })
}

export function submitEventParticipant(eventId, participantChildIds) {
  let createChild = function (childId) {
    return {
      "child_id": childId
    }
  };
  let participantData = {
    "participant": {
    "participant_children_attributes": participantChildIds.map(createChild)
    }
  }
  return Vue.axios.post(`${process.env.BASE_URL_API}/api/events/${eventId}/participants`, participantData)
        .then(res => {
          console.log('SUBMIT EVENT PARTICIPANT SUCCESS')
          console.log(res)
          return res
        })
        .catch(err => {
          console.log('SUBMIT EVENT PARTICIPANT FAILURE')
          console.log(err)
          console.log(Object.entries(err))
          throw err
        })
}

/*
 * UTILS
 */

export function distanceHaversine(lat1, lon1, lat2, lon2) {
        const unit='N' // always return miles
        var radlat1 = Math.PI * lat1/180
        var radlat2 = Math.PI * lat2/180
        var radlon1 = Math.PI * lon1/180
        var radlon2 = Math.PI * lon2/180
        var theta = lon1-lon2
        var radtheta = Math.PI * theta/180
        var dist = Math.sin(radlat1) * Math.sin(radlat2) + Math.cos(radlat1) * Math.cos(radlat2) * Math.cos(radtheta);
        dist = Math.acos(dist)
        dist = dist * 180/Math.PI
        dist = dist * 60 * 1.1515
        if (unit=="K") { dist = dist * 1.609344 }
        if (unit=="N") { dist = dist * 0.8684 }
        return dist.toFixed(1)
    }
    
function capitalize(string) {
  return string.charAt(0).toUpperCase() + string.slice(1);  
}