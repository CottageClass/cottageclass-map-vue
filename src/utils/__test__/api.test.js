import { fetchEvent } from '../api'
import mockAxios from 'axios'

describe('fetchEvent', () => {
  it('has the right data', async () => {
    const event = await fetchEvent('2186')
    expect(mockAxios.get).toHaveBeenCalledTimes(1)
    expect(event.data).toMatchSnapshot()
  })
  it('throws an error with a bad id', async () => {
    await expect(fetchEvent(123123)).rejects.toEqual(Error('failed to fetch event'))
  })
})
