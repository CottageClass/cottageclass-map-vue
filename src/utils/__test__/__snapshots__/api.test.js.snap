// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchEvent has the right data 1`] = `
Object {
  "child": Object {
    "733": Object {
      "attributes": Object {
        "age": 2,
        "parentId": 1184,
      },
      "id": "733",
      "relationships": Object {
        "parent": Object {
          "data": Object {
            "id": "1184",
            "type": "user",
          },
        },
      },
      "type": "child",
    },
    "767": Object {
      "attributes": Object {
        "age": 10,
        "parentId": 1184,
      },
      "id": "767",
      "relationships": Object {
        "parent": Object {
          "data": Object {
            "id": "1184",
            "type": "user",
          },
        },
      },
      "type": "child",
    },
  },
  "event": Object {
    "2186": Object {
      "attributes": Object {
        "activityNames": Array [
          "arts & crafts",
        ],
        "childAgeMaximum": 11,
        "childAgeMinimum": 2,
        "endsAt": "2019-03-23T21:00:00-04:00",
        "foods": Array [
          "tacos ðŸŒ®",
        ],
        "full": true,
        "hasPet": false,
        "hostAdminAreaLevel1": "Massachusetts",
        "hostAdminAreaLevel2": "Worcester County",
        "hostAvatar": null,
        "hostChildAges": Array [
          10,
          2,
        ],
        "hostCreatedAt": "2019-02-20T19:50:54.060Z",
        "hostEmployer": null,
        "hostFacebookUid": "2345507718810193",
        "hostFirstName": "Asa",
        "hostFuzzyLatitude": "42.2621022893375",
        "hostFuzzyLongitude": "-71.7869041541646",
        "hostHighestEducation": null,
        "hostId": 1184,
        "hostImages": Array [],
        "hostInstagramUser": null,
        "hostJobPosition": null,
        "hostLanguages": Array [],
        "hostLinkedinUser": null,
        "hostLocality": "Worcester",
        "hostNeighborhood": "Shrewsbury Street",
        "hostProfileBlurb": null,
        "hostSchool": null,
        "hostSublocality": null,
        "hostTwitterUser": null,
        "hostVerified": false,
        "houseRules": null,
        "maximumChildren": 2,
        "name": "Arts & crafts & tacos ðŸŒ® with Asa",
        "participantsCount": 2,
        "petDescription": null,
        "startsAt": "2019-03-23T18:00:00-04:00",
      },
      "id": "2186",
      "relationships": Object {
        "eventHosts": Object {
          "data": Array [],
        },
        "host": Object {
          "data": Object {
            "id": "1184",
            "type": "user",
          },
        },
        "participants": Object {
          "data": Array [
            Object {
              "id": "227",
              "type": "participant",
            },
            Object {
              "id": "228",
              "type": "participant",
            },
          ],
        },
      },
      "type": "event",
    },
  },
  "participant": Object {
    "227": Object {
      "attributes": Object {
        "eventId": 2186,
        "userAdminAreaLevel1": "Massachusetts",
        "userAdminAreaLevel2": "Worcester County",
        "userAvatar": null,
        "userChildAges": Array [
          3,
        ],
        "userCreatedAt": "2019-02-22T18:56:31.314Z",
        "userEmployer": "City of Worcester",
        "userFacebookUid": "10161426358710224",
        "userFirstName": "Ari",
        "userFuzzyLatitude": "42.2798583744957",
        "userFuzzyLongitude": "-71.7923605922667",
        "userHighestEducation": null,
        "userId": 1230,
        "userImages": Array [
          "https://res.cloudinary.com/cottageclass2/image/upload/v1550865925/avatars/ari_photo_sxw70e.png",
          "https://res.cloudinary.com/cottageclass2/image/upload/v1550865941/avatars/Image_from_iOS_u7dxcz.png",
          "https://res.cloudinary.com/cottageclass2/image/upload/v1550865985/avatars/2013-02-09_13.56.09_zzrtjv.png",
          "https://res.cloudinary.com/cottageclass2/image/upload/v1550865992/avatars/2012-10-19_13.57.39_pzi6hl.png",
          "https://res.cloudinary.com/cottageclass2/image/upload/v1550865999/avatars/2012-06-21_14.44.23_tywfj9.png",
          "https://res.cloudinary.com/cottageclass2/image/upload/v1550866006/avatars/2013-01-26_09.29.01_s1va6c.png",
          "https://res.cloudinary.com/cottageclass2/image/upload/v1550866012/avatars/2012-11-19_21.47.23_lhpttt.png",
          "https://res.cloudinary.com/cottageclass2/image/upload/v1550866024/avatars/2019-01-27_09.18.43_v4c1ui.png",
          "https://res.cloudinary.com/cottageclass2/image/upload/v1550866035/avatars/2013-09-28_21.23.05_tsl6be.png",
        ],
        "userInstagramUser": null,
        "userJobPosition": "Middle School teacher",
        "userLanguages": Array [],
        "userLinkedinUser": null,
        "userLocality": "Worcester",
        "userNeighborhood": "Biotech Park Area",
        "userProfileBlurb": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint.
Roccaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.


Elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint.",
        "userSchool": null,
        "userSublocality": null,
        "userTwitterUser": null,
        "userVerified": false,
      },
      "id": "227",
      "relationships": Object {
        "participantChildren": Object {
          "data": Array [
            Object {
              "id": "208",
              "type": "participantChild",
            },
          ],
        },
      },
      "type": "participant",
    },
    "228": Object {
      "attributes": Object {
        "eventId": 2186,
        "userAdminAreaLevel1": "Massachusetts",
        "userAdminAreaLevel2": "Worcester County",
        "userAvatar": "https://res.cloudinary.com/cottageclass2/image/upload/v1550852508/avatars/ari_photo_qie6qh.png",
        "userChildAges": Array [
          2,
        ],
        "userCreatedAt": "2019-02-22T16:21:51.663Z",
        "userEmployer": null,
        "userFacebookUid": null,
        "userFirstName": "ari",
        "userFuzzyLatitude": "42.2804470048222",
        "userFuzzyLongitude": "-71.7921261738453",
        "userHighestEducation": null,
        "userId": 1228,
        "userImages": Array [],
        "userInstagramUser": null,
        "userJobPosition": null,
        "userLanguages": Array [],
        "userLinkedinUser": null,
        "userLocality": "Worcester",
        "userNeighborhood": "Biotech Park Area",
        "userProfileBlurb": null,
        "userSchool": null,
        "userSublocality": null,
        "userTwitterUser": null,
        "userVerified": false,
      },
      "id": "228",
      "relationships": Object {
        "participantChildren": Object {
          "data": Array [
            Object {
              "id": "209",
              "type": "participantChild",
            },
          ],
        },
      },
      "type": "participant",
    },
  },
  "participantChild": Object {
    "208": Object {
      "attributes": Object {
        "childAge": 3,
        "childId": 768,
      },
      "id": "208",
      "type": "participantChild",
    },
    "209": Object {
      "attributes": Object {
        "childAge": 2,
        "childId": 766,
      },
      "id": "209",
      "type": "participantChild",
    },
  },
  "user": Object {
    "1184": Object {
      "attributes": Object {
        "adminAreaLevel1": "Massachusetts",
        "adminAreaLevel2": "Worcester County",
        "avatar": null,
        "childAges": Array [
          10,
          2,
        ],
        "createdAt": "2019-02-20T19:50:54.060Z",
        "employer": null,
        "facebookUid": "2345507718810193",
        "firstName": "Asa",
        "fuzzyLatitude": "42.2621022893375",
        "fuzzyLongitude": "-71.7869041541646",
        "highestEducation": null,
        "id": 1184,
        "images": Array [],
        "instagramUser": null,
        "jobPosition": null,
        "languages": Array [],
        "linkedinUser": null,
        "locality": "Worcester",
        "neighborhood": "Shrewsbury Street",
        "profileBlurb": null,
        "school": null,
        "sublocality": null,
        "twitterUser": null,
        "verified": false,
      },
      "id": "1184",
      "relationships": Object {
        "children": Object {
          "data": Array [
            Object {
              "id": "767",
              "type": "child",
            },
            Object {
              "id": "733",
              "type": "child",
            },
          ],
        },
      },
      "type": "user",
    },
  },
}
`;
